src/main/java/ogc/rs/apiserver/handlers/UsageLimitEnforcementHandler.java

```diff
@@ -0,0 +1,127 @@
+package ogc.rs.apiserver.handlers;
+
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.RoutingContext;
+import ogc.rs.apiserver.util.AuthInfo;
+import ogc.rs.apiserver.util.OgcException;
+import ogc.rs.database.DatabaseService;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import static ogc.rs.apiserver.handlers.DxTokenAuthenticationHandler.USER_KEY;
+import static ogc.rs.apiserver.util.Constants.*;
+import static ogc.rs.common.Constants.DATABASE_SERVICE_ADDRESS;
+
+/**
+ * This HTTP request handler enforces API usage limits per user and collection.
+ * <p>
+ * Limits are extracted from the user's JWT token and can include either:
+ * - Data usage limit (e.g., "100:mb")
+ * - API hit limit (e.g., 100)
+ * <p>
+ * The limits are applied from the policy issued-at timestamp (iat) defined in the token.
+ */
+public class UsageLimitEnforcementHandler implements Handler<RoutingContext> {
+    private static final Logger LOGGER = LogManager.getLogger(UsageLimitEnforcementHandler.class);
+    Vertx vertx;
+    private final DatabaseService databaseService;
+
+    /**
+     * Constructs the handler with a Vert.x instance and initializes a proxy for DatabaseService.
+     *
+     * @param vertx Vertx instance used to create service proxies.
+     */
+    public UsageLimitEnforcementHandler(Vertx vertx) {
+        this.vertx = vertx;
+        this.databaseService = DatabaseService.createProxy(vertx, DATABASE_SERVICE_ADDRESS);
+    }
+
+    /**
+     * Intercepts each request and checks usage limits based on the user's token.
+     * If the user has exceeded their allowed quota, the request is denied with 429 status.
+     *
+     * @param routingContext Context of the current HTTP request.
+     */
+    @Override
+    public void handle(RoutingContext routingContext) {
+        LOGGER.debug("Usage Limit Enforcement Handler invoked");
+
+        AuthInfo user = routingContext.get(USER_KEY);
+        String userId = user.getUserId().toString();
```

Comment:
Must handle delegate flow, i.e. if it's a delegate token, then the delegator's user ID must be used